### PR TITLE
BUG/TST: fix and test for timezone drop in GroupBy.shift/bfill/ffill

### LINKED ISSUE #19995: Shifting a datetime column with timezone after groupby loses the timezone.
#### Code Sample, a copy-pastable example if possible

```python
weeks = pd.Series(pd.date_range('2018-01', '2018-02', freq='7D', tz='America/New_York'))
week_ago = weeks.groupby([1, 1, 1, 1, 1]).shift()
print(week_ago)
```
Outputs:
```
0                   NaT
1   2018-01-01 05:00:00
2   2018-01-08 05:00:00
3   2018-01-15 05:00:00
4   2018-01-22 05:00:00
dtype: datetime64[ns]
```

#### Problem description

Shifting a groupby'd datetime column removes the timezone. It's not mentioned in the documentation so either the timezone should be preserved or the documentation should mention that timezone is removed (by converting to UTC it seems).

I had a quick look and couldn't find another issue for this. I glanced at the GroupBy shift implementation and my guess is that it's due to cython/numpy not supporting timezones.

I'm happy to work on a fix if someone says what the preferred solution is.

#### Expected Output

```
0                         NaT
1   2018-01-01 00:00:00-05:00
2   2018-01-08 00:00:00-05:00
3   2018-01-15 00:00:00-05:00
4   2018-01-22 00:00:00-05:00
dtype: datetime64[ns, America/New_York]
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-693.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: 3.4.1
pip: 9.0.1
setuptools: 38.5.1
Cython: None
numpy: 1.13.3
scipy: 1.0.0
pyarrow: 0.8.0
xarray: None
IPython: 6.2.1
sphinx: None
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2018.3
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: 2.6.4
feather: 0.4.0
matplotlib: 2.1.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: 1.2.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: 0.1.4
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by noritada**
closes #19995

Timezone info is dropped in GroupBy.shift, bfill, and ffill
since index calculated by Cythonized functions are applied to
NumPy representation of values without timezone.

Could you please find and review this fix?

Many thanks,

-noritada

```
In [1]: import pandas as pd 
   ...: pd.__version__ 
Out[1]: '0.24.2'

In [2]: df = pd.DataFrame([ 
   ...:     ['a', pd.Timestamp('2019-01-01 00:00:00+09:00')], 
   ...:     ['b', pd.Timestamp('2019-01-01 00:05:00+09:00')], 
   ...:     ['a', None], 
   ...:     ['b', None], 
   ...:     ['a', pd.Timestamp('2019-01-01 00:20:00+09:00')], 
   ...:     ['b', pd.Timestamp('2019-01-01 00:25:00+09:00')], 
   ...: ], columns=['id', 'time']) 

In [3]: df['time'].shift(1) 
Out[3]: 
0                         NaT
1   2019-01-01 00:00:00+09:00
2   2019-01-01 00:05:00+09:00
3                         NaT
4                         NaT
5   2019-01-01 00:20:00+09:00
Name: time, dtype: datetime64[ns, pytz.FixedOffset(540)]

In [4]: df.groupby('id')['time'].shift(1) 
Out[4]: 
0                   NaT
1                   NaT
2   2018-12-31 15:00:00
3   2018-12-31 15:05:00
4                   NaT
5                   NaT
Name: time, dtype: datetime64[ns]

In [5]: df.groupby('id')['time'].bfill() 
Out[5]: 
0   2018-12-31 15:00:00
1   2018-12-31 15:05:00
2   2018-12-31 15:20:00
3   2018-12-31 15:25:00
4   2018-12-31 15:20:00
5   2018-12-31 15:25:00
Name: time, dtype: datetime64[ns]

In [6]: df.groupby('id')['time'].ffill() 
Out[6]: 
0   2018-12-31 15:00:00
1   2018-12-31 15:05:00
2   2018-12-31 15:00:00
3   2018-12-31 15:05:00
4   2018-12-31 15:20:00
5   2018-12-31 15:25:00
Name: time, dtype: datetime64[ns]
```

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -120,6 +120,7 @@ Groupby/resample/rolling
 - Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.transform` where applying a timezone conversion lambda function would drop timezone information (:issue:`27496`)
 - Bug in windowing over read-only arrays (:issue:`27766`)
 - Fixed segfault in `pandas.core.groupby.DataFrameGroupBy.quantile` when an invalid quantile was passed (:issue:`27470`)
+- Bug in :meth:`pandas.core.groupby.GroupBy.shift`, :meth:`pandas.core.groupby.GroupBy.bfill` and :meth:`pandas.core.groupby.GroupBy.ffill` where timezone information would be dropped (:issue:`xxxxx`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.1.rst]: If there is not an existing issue for this bug, use the number of this PR.

File: pandas/tests/groupby/test_timezone.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                    Timestamp("2019-01-01 12:00:00", tz="Asia/Tokyo"),
+                    Timestamp("2019-01-01 12:30:00", tz="Asia/Tokyo"),
+                    Timestamp("2019-01-01 14:00:00", tz="Asia/Tokyo"),
+                    Timestamp("2019-01-01 14:30:00", tz="Asia/Tokyo"),
+                ]
+            },
+        ),
+    ],
+)
+def test_shift_bfill_ffill_tz(data, expected_shift, expected_bfill, expected_ffill):
```
Comment (mroeschke) [inline pandas/tests/groupby/test_timezone.py]: Can you put this test in `test_groupby.py`

Also you can parameterize over `shift`, `bfill`, and `ffill`

**Comment by noritada** [comment]
@mroeschke 
Thank you very much for your reviews.
I've updated as per your suggestions.
Could you please check again?

Many thanks,

-noritada

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -1882,3 +1882,69 @@ def test_groupby_axis_1(group_name):
     results = df.groupby(group_name, axis=1).sum()
     expected = df.T.groupby(group_name).sum().T
     assert_frame_equal(results, expected)
+
+
+@pytest.mark.parametrize("tz", [None, "Asia/Tokyo"])
```
Comment (mroeschke) [inline pandas/tests/groupby/test_groupby.py]: We have a `tz_naive_fixture` you can use in the function signature that will parameterize over more timezones.

```
def test_shift_bfill_ffill_tz(tz_naive_fixture, op, expected):
    tz = tz_naive_fixture
    ...
```

**Review by mroeschke — CHANGES_REQUESTED** [review]
Very nice. One small comment otherwise LGTM.

**Comment by noritada** [comment]
@mroeschke
Thank you so much again.
I should have found this function.
Test code updated again.

Many thanks,

-noritada

**Review by mroeschke — APPROVED** [review]

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -2220,26 +2220,28 @@ def _get_cythonized_result(
         base_func = getattr(libgroupby, how)
 
         for name, obj in self._iterate_slices():
+            values = obj._data._values
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py]: Hmm I'm not sure relying on underlying block values is the best way to go about this. Is it possible to just work with the obj here instead?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
Can you update whatsnew for 0.25.2 as well?

**Comment by TomAugspurger** [comment]
1.0.0, unless this is a regression in 0.25.x.

File: doc/source/whatsnew/v0.25.1.rst [hunk 5]
```diff
@@ -120,6 +120,7 @@ Groupby/resample/rolling
 - Bug in :meth:`pandas.core.groupby.DataFrameGroupBy.transform` where applying a timezone conversion lambda function would drop timezone information (:issue:`27496`)
 - Bug in windowing over read-only arrays (:issue:`27766`)
 - Fixed segfault in `pandas.core.groupby.DataFrameGroupBy.quantile` when an invalid quantile was passed (:issue:`27470`)
+- Bug in :meth:`pandas.core.groupby.GroupBy.shift`, :meth:`pandas.core.groupby.GroupBy.bfill` and :meth:`pandas.core.groupby.GroupBy.ffill` where timezone information would be dropped (:issue:`27992`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: can you move to 0.25.2

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. can you do an issue search to see if we have anything about this (IIRC we do). if so, pls reference these issues.

**Comment by jreback** [comment]
also pls merge master

File: pandas/core/groupby/groupby.py (hunk 4 above)
Comment (noritada) [reply pandas/core/groupby/groupby.py]: Sorry for my lack of understanding about underlying block values.
Should I leave the code using `obj.values` except for the last `result = algorithms.take_nd(obj.values, result)`?

**Comment by noritada** [comment]
@WillAyd @TomAugspurger @jreback
Thanks for comments and sorry to be absent for 3 weeks.
I've merged master, move a whatsnew entry to v0.25.2 and document #19995 .

Could you please check this again?

Many thanks,

-noritada

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @noritada

**Comment by TomAugspurger** [comment]
@jreback I think the plan was for 0.25.2 to just be Python 3.8 compat and 0.25 regressions: https://github.com/pandas-dev/pandas/issues/28110. I don't believe this was a regression in 0.25.

**Comment by jreback** [comment]
ok sure np

alright then prob just need to move the actual release note

**Pull request merged**